keel-tagging-plugin/src/main/kotlin/com/netflix/spinnaker/keel/tagging/ResourceTagger.kt

```diff
@@ -0,0 +1,203 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.keel.tagging
+
+import com.netflix.spinnaker.keel.activation.ApplicationDown
+import com.netflix.spinnaker.keel.activation.ApplicationUp
+import com.netflix.spinnaker.keel.actuation.ResourcePersister
+import com.netflix.spinnaker.keel.api.ResourceName
+import com.netflix.spinnaker.keel.api.SPINNAKER_API_V1
+import com.netflix.spinnaker.keel.api.SubmittedResource
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.clouddriver.model.Credential
+import com.netflix.spinnaker.keel.events.DeleteEvent
+import com.netflix.spinnaker.keel.persistence.NoSuchResourceException
+import com.netflix.spinnaker.keel.persistence.ResourceRepository
+import com.netflix.spinnaker.keel.tags.EntityRef
+import com.netflix.spinnaker.keel.tags.EntityTag
+import com.netflix.spinnaker.keel.tags.KEEL_TAG_NAME
+import com.netflix.spinnaker.keel.tags.TagValue
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.context.event.EventListener
+import org.springframework.scheduling.annotation.Scheduled
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+import java.util.concurrent.TimeUnit
+
+/**
+ * A scheduled job that checks to make sure each resource keel is managing
+ * has an entity tag indicating this, as well as linking the entity tag id
+ * to the keel id.
+ *
+ */
+class ResourceTagger(
+  private val resourceRepository: ResourceRepository,
+  private val resourcePersister: ResourcePersister,
+  private val cloudDriverService: CloudDriverService,
+  private val publisher: ApplicationEventPublisher
+) {
+  private val log = LoggerFactory.getLogger(javaClass)
+
+  private var enabled = false
+  private var accounts: Set<Credential> = emptySet()
+  private var accountsUpdateTimeS = 0L
+  private var accountsUpdateFrequencyS = TimeUnit.MINUTES.toSeconds(10)
+  private val transforms = mapOf(
+    "ec2" to "aws"
+  )
+
+  init {
+    syncAccounts()
+  }
+
+  @EventListener(ApplicationUp::class)
+  fun onApplicationUp() {
+    log.info("Application up, enabling scheduled resource tagging")
+    enabled = true
+  }
+
+  @EventListener(ApplicationDown::class)
+  fun onApplicationDown() {
+    log.info("Application down, disabling scheduled resource tagging")
+    enabled = false
+  }
+
+  @Scheduled(fixedDelayString = "\${keel.resource-tagger.frequency:PT10S}")
+  fun checkResources() {
```

Comment:
this looks like it will handle the case of ensuring a tag exists for every resource in keel

Do we also want to ensure a resource exists in keel for every tag?

Reply:
refactored to do things on create and delete instead of this polling. 

That might be a good future check. Can you think of any cases where this might bite us?